src/EventStore.ClientAPI.Embedded/EmbeddedEventStorePersistentSubscription.cs

```diff
@@ -10,7 +10,7 @@ internal class EmbeddedEventStorePersistentSubscription : EventStorePersistentSu
 
         internal EmbeddedEventStorePersistentSubscription(
             string subscriptionId, string streamId,
-            Action<EventStorePersistentSubscriptionBase, ResolvedEvent> eventAppeared,
+            Func<EventStorePersistentSubscriptionBase, ResolvedEvent, Task> eventAppeared,
```

Comment:
I think you should consider passing in a cancellation token
`Func<EventStorePersistentSubscriptionBase, ResolvedEvent, CancellationToken, Task>`

That way if a subscription is disposed (/system shutting down), `eventAppeared` handlers can participate in co-operative cancellation. Usually that means passing the token to something else that does async I/O and is cancellable.

This is the signature I used https://github.com/SQLStreamStore/SQLStreamStore/blob/master/src/SqlStreamStore/Subscriptions/StreamMessageReceived.cs#L20

Also something that may or may not be useful, [my TaskQueue implementation](https://github.com/SQLStreamStore/SQLStreamStore/blob/master/src/SqlStreamStore/Infrastructure/TaskQueue.cs).